include/boost/math/ccmath/fma.hpp

```diff
@@ -0,0 +1,128 @@
+//  (C) Copyright Matt Borland 2022.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_CCMATH_FMA_HPP
+#define BOOST_MATH_CCMATH_FMA_HPP
+
+#include <cmath>
+#include <limits>
+#include <type_traits>
+#include <boost/math/tools/is_constant_evaluated.hpp>
+#include <boost/math/ccmath/isinf.hpp>
+#include <boost/math/ccmath/isnan.hpp>
+
+namespace boost::math::ccmath {
+
+namespace detail {
+
+template <typename T>
+constexpr T fma_imp(const T x, const T y, const T z) noexcept
+{
+    #if defined(__GNUC__) && !defined(__clang__) && !defined(__INTEL_COMPILER) && !defined(__INTEL_LLVM_COMPILER)
+    if constexpr (std::is_same_v<T, float>)
+    {
+        return __builtin_fmaf(x, y, z);
+    }
+    else if constexpr (std::is_same_v<T, double>)
+    {
+        return __builtin_fma(x, y, z);
+    }
+    else if constexpr (std::is_same_v<T, long double>)
+    {
+        return __builtin_fmal(x, y, z);
+    }
+    #endif
+    
+    // If we can't use compiler intrinsics hope that -fma flag optimizes this call to fma instruction
+    return (x * y) + z;
+}
+
+} // Namespace detail
+
+template <typename Real, std::enable_if_t<!std::is_integral_v<Real>, bool> = true>
+constexpr Real fma(Real x, Real y, Real z) noexcept
+{
+    if (BOOST_MATH_IS_CONSTANT_EVALUATED(x))
+    {
+        if (x == 0 && boost::math::ccmath::isinf(y))
```

Comment:
Is this not a bit too branchy here? I worry about the case where x is constant evaluated but y and z are runtime and this becomes slow as a result.

Reply:
x is only a place holder for old versions (GCC 6-8) of gcc's builtin constant evaluated. On GCC 9+ and MSVC this compiles down to `std::is_constant_evaluated()`. Either the whole thing can be constant evaluated or none of it.